### PR TITLE
Do not scroll chat view in narrow mode when chat list is displayed #1040

### PR DESCRIPTION
**Pull request by crui3er**

### CODE REVIEW

File: src/dotnet/UI.Blazor/Services/PanelsUI/LeftPanel.cs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+        bool oldIsVisible;
+        lock (_lock) {
+            oldIsVisible = _isVisible.Value;
+            if (oldIsVisible != value)
+                _isVisible.Value = value;
+        }
+        if (oldIsVisible != value) {
+            Log.LogDebug("Visibility changed: {IsVisible}", value);
+            Owner.History.Save<OwnHistoryState>();
```
Comment (FrolYo) [inline src/dotnet/UI.Blazor/Services/PanelsUI/LeftPanel.cs:40]: due to lock above you expect race conditions in this method.
On the other hand here you save to history state IsVisible.Value in a not concurrent safe manner. At this line you can safe state which is already modified by a concurrent call.
i.e. you can save `false, false` when you must save `false, true `

File: src/dotnet/UI.Blazor/Services/PanelsUI/MiddlePanel.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+            ComputeIsVisible);
+    }
+
+    private async Task<bool> ComputeIsVisible(IComputedState<bool> state, CancellationToken cancellationToken)
+    {
+        var screenSize = await Owner.ScreenSize.Use(cancellationToken);
+        var isWide = screenSize.IsWide();
+        if (isWide)
+            return true;
+        if (await Owner.Left.IsVisible.Use(cancellationToken))
```
Comment (FrolYo) [inline src/dotnet/UI.Blazor/Services/PanelsUI/MiddlePanel.cs:28]: configure await

File: src/dotnet/UI.Blazor/Services/SwitchableDelayer.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
+namespace ActualChat.UI.Blazor.Services;
+
+public class SwitchableDelayer
+{
+    private readonly object _syncObject = new ();
```
Comment (FrolYo) [inline src/dotnet/UI.Blazor/Services/SwitchableDelayer.cs:5]: afaik we decided to name it _lock

File: src/dotnet/UI.Blazor/Services/SwitchableDelayer.cs [hunk 4]
```diff
@@ -0,0 +1,33 @@
+namespace ActualChat.UI.Blazor.Services;
+
+public class SwitchableDelayer
```
Comment (FrolYo) [inline src/dotnet/UI.Blazor/Services/SwitchableDelayer.cs:3]: isn't it just a sort of reset event/semaphor?

**Review by FrolYo — CHANGES_REQUESTED** [review]

File: src/dotnet/UI.Blazor/Services/PanelsUI/MiddlePanel.cs (hunk 2 above)
Comment (crui3er) [reply src/dotnet/UI.Blazor/Services/PanelsUI/MiddlePanel.cs:28]: added

File: src/dotnet/UI.Blazor/Services/SwitchableDelayer.cs (hunk 3 above)
Comment (crui3er) [reply src/dotnet/UI.Blazor/Services/SwitchableDelayer.cs:5]: fixed

File: src/dotnet/UI.Blazor/Services/PanelsUI/LeftPanel.cs (hunk 1 above)
Comment (crui3er) [reply src/dotnet/UI.Blazor/Services/PanelsUI/LeftPanel.cs:40]: kept it without changes

File: src/dotnet/Chat.UI.Blazor/Components/LeftPanel/LeftPanel.razor [hunk 5]
```diff
@@ -4,22 +4,22 @@
...
+    VisibilityChanged="@(isOpen => Panels.Left.SetIsVisible(isOpen))">
     <div class="flex-x h-full bg-04">
         <LeftPanelButtons/>
         <LeftPanelContent/>
     </div>
 </SideNav>
 
 @code {
-    [Inject] private NavbarUI NavbarUI { get; init; } = null!;
+    [Inject] private PanelsUI Panels { get; init; } = null!;
```
[the remark this answers is not in the document]
Comment (crui3er) [reply src/dotnet/Chat.UI.Blazor/Components/LeftPanel/LeftPanel.razor:16]: fixed

**Review by FrolYo — APPROVED** [review]

**Pull request merged**